cdm/s3/src/main/java/thredds/inventory/s3/MFileS3.java

```diff
@@ -285,13 +298,44 @@ public boolean exists() {
     return exists;
   }
 
-  // Update file exists by fetching from a head request
   private void updateExists() {
+    if (isDirectory()) {
+      exists = directoryExists();
+    } else {
+      exists = key == null ? bucketExists() : objectExists();
+    }
+  }
+
+  private boolean directoryExists() {
+    final S3Client client = getClient();
+    if (client == null) {
+      return false;
+    }
+
+    try {
+      final ListObjectsV2Response listObjects =
+          client.listObjectsV2(ListObjectsV2Request.builder().bucket(cdmS3Uri.getBucket()).prefix(key).build());
+      return !listObjects.contents().isEmpty();
+    } catch (NoSuchBucketException e) {
+      return false;
+    }
+  }
+
+  private boolean bucketExists() {
+    try {
+      getHeadBucketResponse();
+      return true;
+    } catch (NoSuchBucketException e) {
```

Comment:
Sorry if this is just my bronchitis brain, but does this method throw that error? I thought it just returns null.

Reply:
Yes it does-- the `getHeadBucketResponse` does a `headBucket` which has this signature: 
```
default HeadBucketResponse headBucket(HeadBucketRequest headBucketRequest) throws NoSuchBucketException, AwsServiceException,
            SdkClientException, S3Exception
```
But the `NoSuchBucketException` is a `RuntimeException` so not declared for `getHeadBucketResponse` so it's not obvious. The test `shouldReturnFalseForNonExistentBucket` would fail without this catch statement.

Reply:
but maybe we should also test the response code? Maybe there could be a bucket you don't have access to so it doesn't throw `NoSuchBucketExceptions`?

Reply:
I think at least checking for null would be good

Reply:
Added null checks and response code checks in bf0a581e759eed0c7031cfe178a284cd8c4f596a